### PR TITLE
(proof-new) New proofs in arith::Constraint::externalExplain

### PR DESCRIPTION
**Pull request by alex-ozdemir**
This commit threads proofs through the core of arith:
Constraint::externalExplain, which recursively explain arith literals.
One of the base cases here is asking the EE for an explanation, through
the congruence manager. To delay implementing proofs in
ArithCongruenceManager to a separate commit, we stub out proof
production in ArithCongruenceManager::explain for now.

### CODE REVIEW

File: src/theory/arith/congruence_manager.cpp [hunk 1]
```diff
@@ -375,11 +375,22 @@ Node ArithCongruenceManager::explainInternal(TNode internal){
...
-  return explainInternal(internal);
+  Node exp = explainInternal(internal);
+  if (isProofEnabled())
+  {
+    Node impl = NodeManager::currentNM()->mkNode(Kind::IMPLIES, exp, external);
+    // For now, we just trust
+    auto pfOfImpl = d_pnm->mkNode(PfRule::INT_TRUST, {}, { impl });
+    return d_pfGenExplain->mkTrustNode(impl, pfOfImpl);
+  }
+  else
```
Comment (ajreynol) [inline src/theory/arith/congruence_manager.cpp:390]: Remove redundant "else" after returns.

File: src/theory/arith/constraint.cpp [hunk 2]
```diff
@@ -1522,68 +1519,203 @@ Node Constraint::externalExplain(const ConstraintCPVec& v, AssertionOrder order)
...
   }
+  Assert(hasProof());
+  Assert(!isAssumption() || assertedToTheTheory());
+  Assert(!isInternalAssumption());
+  std::shared_ptr<ProofNode> pf{};
 
   if(assertedBefore(order)){
+    Debug("pf::arith::explain") << "  already asserted" << std::endl;
     nb << getWitness();
+    if (options::proofNew())
```
Comment (ajreynol) [inline src/theory/arith/constraint.cpp:1547]: It's preferred not to check options during solving, can you update this based on checking e.g. if d_database has proofs enabled (which is the case if its proof node manager is non-null)?

File: src/theory/arith/constraint.cpp [hunk 3]
```diff
@@ -1522,68 +1519,203 @@ Node Constraint::externalExplain(const ConstraintCPVec& v, AssertionOrder order)
...
+      {
+        pf = d_database->d_pnm->mkNode(
+            PfRule::MACRO_SR_PRED_TRANSFORM, {pf}, {getProofLiteral()});
+      }
+    }
   }else if(hasEqualityEngineProof()){
-    d_database->eeExplain(this, nb);
+    Debug("pf::arith::explain") << "  going to ee:" << std::endl;
+    TrustNode exp = d_database->eeExplain(this);
+    if (options::proofNew())
```
Comment (ajreynol) [inline src/theory/arith/constraint.cpp:1561]: Ditto.

File: src/theory/arith/constraint.cpp [hunk 4]
```diff
@@ -1522,68 +1519,203 @@ Node Constraint::externalExplain(const ConstraintCPVec& v, AssertionOrder order)
...
     Assert(!isAssumption());
     AntecedentId p = getEndAntecedent();
     ConstraintCP antecedent = d_database->d_antecedents[p];
+    std::vector<std::shared_ptr<ProofNode>> children;
 
     while(antecedent != NullConstraint){
       Debug("pf::arith::explain") << "Explain " << antecedent << std::endl;
-      antecedent->externalExplain(nb, order);
+      auto pn = antecedent->externalExplain(nb, order);
+      if (options::proofNew())
```
Comment (ajreynol) [inline src/theory/arith/constraint.cpp:1604]: Ditto.

File: src/theory/arith/constraint.cpp [hunk 5]
```diff
@@ -1522,68 +1519,203 @@ Node Constraint::externalExplain(const ConstraintCPVec& v, AssertionOrder order)
...
-      NodeBuilder<> nb(kind::AND);
-      Assert(!isAssumption());
-
-      AntecedentId p = listEnd;
-      ConstraintCP antecedent = d_database->d_antecedents[p];
-      while(antecedent != NullConstraint){
-        antecedent->externalExplain(nb, order);
-        --p;
-        antecedent = d_database->d_antecedents[p];
+    if (options::proofNew())
```
Comment (ajreynol) [inline src/theory/arith/constraint.cpp:1612]: Ditto.

File: src/theory/arith/constraint.cpp [hunk 6]
```diff
@@ -1522,68 +1519,203 @@ Node Constraint::externalExplain(const ConstraintCPVec& v, AssertionOrder order)
...
+          // Enumerate d_farkasCoefficients as nodes.
+          std::vector<Node> farkasCoeffs;
+          for (Rational r : *getFarkasCoefficients())
+          {
+            farkasCoeffs.push_back(nm->mkConst<Rational>(r));
+          }
+
+          // Apply the scaled-sum rule.
+          std::shared_ptr<ProofNode> sumPf =
+              d_database->d_pnm->mkNode(PfRule::ARITH_SCALE_SUM_UPPER_BOUNDS,
```
Comment (ajreynol) [inline src/theory/arith/constraint.cpp:1649]: Nit: the code throughout would be a bit cleaner if you cached `ProofNodeManager * pnm = d_database->d_pnm`.

File: src/theory/arith/congruence_manager.cpp [hunk 7]
```diff
@@ -474,6 +485,26 @@ void ArithCongruenceManager::addSharedTerm(Node x){
...
+  auto nm = NodeManager::currentNM();
+  if (an == nm->mkConst(true))
+  {
+    return {};
+  }
+  else if (an.getKind() != Kind::AND)
+  {
+    return {an};
+  }
+  else
```
Comment (ajreynol) [inline src/theory/arith/congruence_manager.cpp:499]: Redundant else.

**Review by ajreynol — CHANGES_REQUESTED** [review]
A few minor style-related requests.

**Comment by alex-ozdemir** [comment]
Thanks for the review, Andy. I think I've addressed most of your comments. Let me know if you see further possible improvements!

File: src/theory/arith/constraint.cpp [hunk 8]
```diff
@@ -1706,14 +1843,10 @@ ConstraintP ConstraintDatabase::getBestImpliedBound(ArithVar v, ConstraintType t
     }
   }
 }
-Node ConstraintDatabase::eeExplain(const Constraint* const c) const{
-  Assert(c->hasLiteral());
-  return d_congruenceManager.explain(c->getLiteral());
-}
 
-void ConstraintDatabase::eeExplain(const Constraint* const c, NodeBuilder<>& nb) const{
+TrustNode ConstraintDatabase::eeExplain(const Constraint* const c) const{
```
Comment (ajreynol) [inline src/theory/arith/constraint.cpp:1847]: Remember to format.

**Review by ajreynol — APPROVED** [review]
LGTM, remember to format.

File: src/theory/arith/constraint.cpp (hunk 8 above)
Comment (alex-ozdemir) [reply src/theory/arith/constraint.cpp]: Whoops. Thanks!

**Pull request merged**